### ISSUE TITLE
RStudio's visual markdown editor stays loading forever

### ISSUE THREAD

**Issue by kotatsuyaki** [NixOS/nixpkgs#210826]
### Describe the bug
When trying to switch to the "visual editor" for Quarto markdown files, the editor stays loading with a spinning indicator in the middle forever (20 minutes+), without any error messages both on the GUI and on stdout/stderr. See the screenshot below.

### Steps To Reproduce
1. Run RStudio by `nix run nixpkgs#rstudio`, where my `nixpkgs` is pinned to commit [`b3818a4`](https://github.com/NixOS/nixpkgs/tree/b3818a46e686f24561a28eaa9fcf35e18b8d8e89).
2. Create a Quarto `.qmd` file by clicking `New Blank File` => `Quarto Doc...` in the `Files` panel, and insert the following Markdown content.

   ```markdown
   ---
   title: Test Document
   ---
   
   # Test Heading
   
   ## Test Heading
   
   Test body
   ```
3. On the top-left corner of the editor panel, click `Visual` to switch from the source editor to the visual editor.

### Expected behavior
The visual editor should be able to render the content and provide editing functionalities.

### Screenshots
![Screenshot_20230115_125814](https://user-images.githubusercontent.com/17219127/212523662-840470d6-2122-4632-8e8e-9eb616ba3bc3.png)

### Additional context
rstudio/rstudio#10674 - This closed issue in the RStudio repository suggests that the problem may have something to do with the `RSTUDIO_PANDOC` environment variable.  On my RStudio, the value of this environment variable is a path under RStudio in `/nix/store`:

```R
> Sys.getenv('RSTUDIO_PANDOC')
[1] "/nix/store/1fdbg2jzmmvysbrq3dck04zsjflaln0s-RStudio-2022.07.1+554/lib/rstudio/bin/quarto/bin/tools"
```

However, this path does not exist:

```console
$ ls /nix/store/1fdbg2jzmmvysbrq3dck04zsjflaln0s-RStudio-2022.07.1+554/lib/rstudio/bin/quarto/bin/tools
ls: cannot access '/nix/store/1fdbg2jzmmvysbrq3dck04zsjflaln0s-RStudio-2022.07.1+554/lib/rstudio/bin/quarto/bin/tools': No such file or directory
```

This seems odd, since the Nix derivation for RStudio already [patches the path to be `"@pandoc@"`](https://github.com/NixOS/nixpkgs/blob/b3818a46e686f24561a28eaa9fcf35e18b8d8e89/pkgs/applications/editors/rstudio/pandoc-nix-path.patch) and [uses `substituteInPlace` to replace `@pandoc@`](https://github.com/NixOS/nixpkgs/blob/b3818a46e686f24561a28eaa9fcf35e18b8d8e89/pkgs/applications/editors/rstudio/default.nix#L149) with a path to the `pandoc` binary from Nixpkgs.

### Notify maintainers
@ciil and @cfhammill 

### Metadata

`nixpkgs` pinned in the registry.

```console
$ nix registry list | grep nixpkgs
system flake:nixpkgs path:/nix/store/d02gi2ddx87l037ghfpxg8z7nxkd6f7n-source?lastModified=1673027386&narHash=sha256-Wjt+oDhRLcf3opIjUKHGN+Xrp3w2ZDms6bO4pCLvsco=&rev=b3818a46e686f24561a28eaa9fcf35e18b8d8e89
```

Information other than `nixpkgs` version (which is inaccurate due to me using Flakes) obtained by `nix-info`.

```console
$ nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 6.0.17, NixOS, 23.05 (Stoat), 23.05.20230106.b3818a4`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.12.0`
```

**Comment by hervyqa** [comment]
i confirm it.

RStudio 2022.07.1+554 "Spotted Wakerobin" Release (99999999999999999999999999999999, 1980-01-01) for Linux
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.15.13 Chrome/87.0.4280.144 Safari/537.36

 - system: `"x86_64-linux"`
 - host os: `Linux 6.3.5, NixOS, 23.05 (Stoat), 23.05.419.3a70dd92993`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - channels(hervyqa): `""`
 - channels(root): `"nixos-23.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
 
![Screenshot_20230606_071730](https://github.com/NixOS/nixpkgs/assets/45872139/7bbe0d3b-3097-4480-a3e5-9f24aa459ed9)

**Comment by kevinushey** [comment]
I can try to help. The main thing that stands out for this issue -- RStudio now bundles [Quarto](https://quarto.org/docs/download/), and uses the version of Pandoc bundled with Quarto -- hence the new path seen for `RSTUDIO_PANDOC`.

Perhaps the key bit is that the default Pandoc path is now read from here:

https://github.com/rstudio/rstudio/blob/7afe4f1a25b8426727c92ec7345e60c311238931/src/cpp/session/session-options.json#L713-L719

and also see how the Pandoc path is set here:

https://github.com/rstudio/rstudio/blob/7afe4f1a25b8426727c92ec7345e60c311238931/src/cpp/session/include/session/SessionConstants.hpp#L144-L148

Based on that, and given the patch in https://github.com/NixOS/nixpkgs/blob/b3818a46e686f24561a28eaa9fcf35e18b8d8e89/pkgs/applications/editors/rstudio/pandoc-nix-path.patch, I think both of the Pandoc path constants need to be patched.

**Comment by nviets** [comment]
Hi @kevinushey - thanks so much for your quick reply and for taking a look. This issue has been open awhile, and I should have included a link to the master branch. @Kupac's just update to [Rstudio 2023 09 0 463](https://github.com/NixOS/nixpkgs/pull/252047), and I think he addresses some of what you pointed out. @Kupac does your patch adjust for both of the Pandoc path constants? Latest patch from master: [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/rstudio/pandoc-nix-path.patch).

We might move this discussion to a branch soon to get some build logs going.

**Comment by kevinushey** [comment]
The other thing you might try, to get a bit more information -- if you right click somewhere in the IDE surface, you should get Inspect Element as a menu option; click that to open the RStudio developer tools. If you then try to open the visual editor while those tools are up, you might get some extra information in the console there regarding what's going wrong. 

You can also click on the Source tab, and try to stop on exception, to see if you can get some useful backtraces -- unfortunately the JS is likely to be obfuscated so it may be challenging to divine the issue just form that.

**Comment by Kupac** [comment]
Thanks, this is great advice to troubleshoot the issue. I'll check the patches first, I don't think I patched the json file, I think we missed that it's referred to in two places

**Comment by biscotty666** [comment]
Has any progress been made on this? Ty.

**Comment by b-rodrigues** [comment]
A new Rstudio got released, it'll be a good opportunity to revisit visual mode https://docs.posit.co/ide/news/

**Comment by biscotty666** [comment]
Cool. I'm happy to test it or otherwise help if I can.

⁣Get BlueMail for Android ​

**Comment by 6d6f6d656e746f** [comment]
Did someone end up with a solution?

**Comment by biscotty666** [comment]
No, and it's really frustrating. I guess people who use R don't use NixOS much. The best thing I've come up with is running it in a Fedora distrobox. I can't seem to do R Markdown documents but Quarto documents work. I haven't been doing anything in R recently, but when I return to it I may just dual-boot :-(

**Comment by Kupac** [comment]
Indeed, thanks to @TomaSajt restructuring the whole package, this also got fixed. The issue can be closed now.

**Comment by b-rodrigues** [comment]
The latest RStudio is an electron app, and indeed, visual editor now works thanks to @TomaSajt . Just for future reference, if you're using Nix on Ubuntu and try to run RStudio, you'll likely end up with an error message. This is due to how Apparmor is set up. 

You will need to create an RStudio-specific AppArmor profile. To do so create this apparmor profile:

```bash
sudo nano /etc/apparmor.d/nix.rstudio
```

Populate it with:

```
profile nix.rstudio /nix/store/*-RStudio-*-wrapper/bin/rstudio flags=(unconfined) {
    userns,
}
```

Save it, load the profile and start RStudio:

```bash
sudo apparmor_parser -r /etc/apparmor.d/nix.rstudio
sudo systemctl reload apparmor
```

**Comment by biscotty666** [comment]
It's actually been working for me for some time now 😄, so I thought this was already addressed. Thanks so much.
